sdk/core/System.ClientModel/src/ModelReaderWriter/ModelReaderWriter.cs

```diff
@@ -91,32 +157,103 @@ public static BinaryData Write(object model, ModelReaderWriterOptions? options =
+    /// <summary>
+    /// Converts the <see cref="BinaryData"/> into a <typeparamref name="T"/>.
+    /// </summary>
+    /// <param name="data">The <see cref="BinaryData"/> to convert.</param>
+    /// <param name="context"> The <see cref="ModelReaderWriterContext"/> to use.</param>
+    /// <param name="options">The <see cref="ModelReaderWriterOptions"/> to use.</param>
+    /// <returns>A <typeparamref name="T"/> representation of the <see cref="BinaryData"/>.</returns>
+    /// <exception cref="InvalidOperationException">Throws if <typeparamref name="T"/> does not have a public or internal parameterless constructor.</exception>
+    /// <exception cref="FormatException">If the model does not support the requested <see cref="ModelReaderWriterOptions.Format"/>.</exception>
+    /// <exception cref="ArgumentNullException">If <paramref name="data"/> is null.</exception>
+    /// <exception cref="MissingMethodException">If <typeparamref name="T"/> does not have a public or non public empty constructor.</exception>
+    public static T? Read<T>(
+        BinaryData data,
+        ModelReaderWriterContext context,
+        ModelReaderWriterOptions? options = default)
+    {
+        return ReadInternal<T>(
+            data,
+            context,
+            options);
+    }
 
-        return GetInstance<T>().Create(data, options);
+    /// <summary>
+    /// Converts the <see cref="BinaryData"/> into a <paramref name="returnType"/>.
+    /// </summary>
+    /// <param name="data">The <see cref="BinaryData"/> to convert.</param>
+    /// <param name="returnType">The type of the object to convert and return.</param>
+    /// <param name="options">The <see cref="ModelReaderWriterOptions"/> to use.</param>
+    /// <returns>A <paramref name="returnType"/> representation of the <see cref="BinaryData"/>.</returns>
+    /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not implement <see cref="IPersistableModel{T}"/>.</exception>
+    /// <exception cref="InvalidOperationException">Throws if <paramref name="returnType"/> does not have a public or internal parameterless constructor.</exception>
+    /// <exception cref="FormatException">If the model does not support the requested <see cref="ModelReaderWriterOptions.Format"/>.</exception>
+    /// <exception cref="ArgumentNullException">If <paramref name="data"/> or <paramref name="returnType"/> are null.</exception>
+    /// <exception cref="MissingMethodException">If <paramref name="returnType"/> does not have a public or non public empty constructor.</exception>
+    public static object? Read(
+        BinaryData data,
+        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] Type returnType,
+        ModelReaderWriterOptions? options = default)
+    {
+        return ReadInternal(
+            data,
+            returnType,
+            s_reflectionContext.Value,
```

Comment:
This method needs to be marked `RequiresUnreferencedCode` because it uses a `ReflectionContext`, which requires unreferenced code.